# Hello Python!

Here, we'll take some first steps with the Python programming language. This tutorial starts from scratch so, if you've never written a line of code in your life, you've come to the right place!

### What we're aiming to do

We often find that knowing just a little bit of programming can get you a long way. We're aiming to teach you enought that you know how to:

1. Identify problems that you can solve with a bit of programming
2. Write and execute simple scripts
3. Know how to find help when you get stuck
    
### What we're not doing

This is not a comprehensive programming course! You're not going to be Mr. Robot or Neo. You won't learn every trick in the book, but hopefully just enough to get started.

### Aren't there lots of these tutorials on the internet?

Absolutely! Go check them out! Some recommendations are below:
- [Programiz](https://www.programiz.com/python-programming)
- [Python practice book](https://anandology.com/python-practice-book/index.html)
- [Full stack python](https://www.fullstackpython.com/)

# Getting started

If you've never written any Python at all before, the first thing we'll have to do is install Python itself. Yes, programming languages have to be installed just like anything else. 

To do this, go to: https://www.python.org/

Hover over the 'Download' menu. The web site should detect the operating system you're using and offer up the correct package. You're looking for Python 3.10.

Download the package and install using the installation wizard.

<img src="img/download.PNG" />

Once Python is installed, go to your start menu. In the 'Recently added' section, you should see **IDLE (Python 3.10 64-bit)**. Start this application.

<img src="img/run_idle.PNG" />

We should see something like this:

<img src="img/idle_start.PNG" />

This is what we call a Python 'interpreter'. It 'interprets' the code we write in Python and converts it to something that computers understand (0's and 1's).

Let's write some code!

# Python basic data types

Python has three basic data types:

- Integer (referred to as int): whole numbers, either positive or negative or zero
- Float (referred to as float): decimal numbers, either positive or negative or zero
- String (referred to as str): text (can include numbers)

We'll mostly be working with strings and integers today so let's have a look at these.

## Strings

A string in python is a collection of words or other characters. These can include letters, numbers, spaces or other special characters. Strings are denoted with either single ```''``` or double quotes ```""```

Below are some properties of strings:

Strings can be added together
```
>>> x = "hello"
>>> x += ' everyone'
>>> x
hello everyone
```

Strings can be tested for other strings
```
>>> 'hell' in x
True
```

Strings can be subsetted by an index
```
>>> x[0:10]
'hello ever'
```
Strings can be split on a given character (here, on a space ```' '```)
```
>>> x.split(' ')
['hello', 'everyone']
```

## Integers

This one is straight-forward.
```
>>> y = 100
>>> y += 10
>>> y
110
```
```
>>> z = 100
>>> z -= 65
>>> z
35
```
```
>>> a = 4
>>> b = 2
>>> a * b
8
```
```
>>> a / b
2.0
```
Note that, for division, the result is a float (decimal number). This behaviour is different between Python 2 and Python 3 (but we'll stick with 3 so don't worry!)

# Python basic data structures

Data structures refer to different ways that we can organise and access our data (strings, integers, floats etc). Python has many different data structures but you can go a long way with programming by understanding just a few:

- Lists
- Tuples
- Dictionaries
- Sets

We'll mostly use lists today so let's look at those.

## Lists

Lists store multiple different items in a single varialb. Similar to how we think about lists in real life, lists have an order, can store different data types (strings, integers, floats...) and can be added to and removed from.

Lists are declared in python with square brackets: ```[]```

Below are some examples of how we can work with lists:

We can append items to a list
```
>>> x = []
>>> x.append(1)
>>> x.append(2)
>>> x
[1, 2]
```

Items in a list can be accessed using an index (starting from 0!)
```
>>> x = ['a','b','c','d','e','f']
>>> x[0]
'a'
>>> x[2]
'c'
>>> x[-1]
'f'
```
We can select ranges from within a list:
```
>>> x[2:4]
['c', 'd']
```
Lists have a length:
```
>>> x = ['a','b','c','d','e','f']
>>> len(x)
6
```

Lists can contain mixed data types:
```
>>> x = ['a','b',100,'d','e','f']
```
Lists can contain other lists:
```
>>> x = ['a','b','c', [1, 2, 3],'e','f']
```

We cam 'iterate' through lists:
```
>>> for element in x:
...     print(element)
... 
a
b
c
[1, 2, 3]
e
f
```


## Scripts

Above we were typing code into the interpreter line-by-line and looking at the results. This is fine but will quickly becomes tedious as we write more and more code. We'd be better off writing all our code into a file (a ```script```) and then telling python to run all the code at once.

- Go to 'File' and and select 'New File'
- A new window will appear - This is your script file!
- Here, you can type some code
- To execute your code, go to 'Run' and select 'Run Module"
- You'll be prompted to save your file and then the code will run in the interpreter window

<img src="img/script.PNG" width="75%" />


## Working with files

We frequently need to use python to open a file on our computers and work with the contents. For the rest of the tutorial, we'll learn how to read a file, iterate through lines of the file and conduct various filterings and comparisons.

The ```fasta``` directory contains the files we'll use in this tutorial. Our first job is to look at the data files and see what we're dealing with. Open one the file called ```small.fasta``` in a text editor (notepad or similar).

<img src="img/fasta.PNG" width="95%" />

The data is in the [FASTA file format](https://en.wikipedia.org/wiki/FASTA_format) which is common for storing protein sequences. Here, each protein entry begins with a 'header' line that begins with ```>``` followed by some text that can provide database identifiers, protein names and descriptions, taxonomy information and so on.

The lines following header line provide the sequence of the protein.

### Opening a file

In the 'fasta' directory, we have a few ```.fasta``` files. We'll start by reading the ```small.fasta``` file

In [8]:
# open the FASTA file in 'read' mode
fasta_file = open('fasta/small.fasta','r').readlines()

# iterate through lines
for line in fasta_file:
    print(line.strip())

>sp|Q9ULX7|CAH14_HUMAN Carbonic anhydrase 14 OS=Homo sapiens OX=9606 GN=CA14 PE=1 SV=1
MLFSALLLEVIWILAADGGQHWTYEGPHGQDHWPASYPECGNNAQSPIDIQTDSVTFDPD
LPALQPHGYDQPGTEPLDLHNNGHTVQLSLPSTLYLGGLPRKYVAAQLHLHWGQKGSPGG
SEHQINSEATFAELHIVHYDSDSYDSLSEAAERPQGLAVLGILIEVGETKNIAYEHILSH
LHEVRHKDQKTSVPPFNLRELLPKQLGQYFRYNGSLTTPPCYQSVLWTVFYRRSQISMEQ
LEKLQGTLFSTEEEPSKLLVQNYRALQPLNQRMVFASFIQAGSSYTTGEMLSLGVGILVG
CLCLLLAVYFIARKIRKKRLENRKSVVFTSAQATTEA


Let's explain the above:

 - On line 2, we use 'open' to access our file
     - the first argument ```fasta/small.fasta``` is the path to the file we want to access
     - the second argument ```r``` tells python to open the file in 'read' mode
     - the ```.readlines()``` call tells python to separate the file into individual lines
 - This creates a ```list``` containing the text from our file. On link 5, we iterate through the list in the same way that we learned earlier
 - On line 6, we print the line
     - The ```.strip()``` call removes extra spaces from the start and end of the text
         - See what happens if you remove the ```.strip()``` call and re-run the code 

Let's add a filter in our ```for``` loop so we print only the FASTA header line

In [12]:
# open the FASTA file in 'read' mode
fasta_file = open('fasta/small.fasta','r').readlines()

# iterate through lines
for line in fasta_file:
    if '>' in line:
        print(line.strip())

>sp|Q9ULX7|CAH14_HUMAN Carbonic anhydrase 14 OS=Homo sapiens OX=9606 GN=CA14 PE=1 SV=1


What if we wanted to print the sequence and not the header lines?

In [13]:
# open the FASTA file in 'read' mode
fasta_file = open('fasta/small.fasta','r').readlines()

# iterate through lines
for line in fasta_file:
    if '>' not in line:
        print(line.strip())

MLFSALLLEVIWILAADGGQHWTYEGPHGQDHWPASYPECGNNAQSPIDIQTDSVTFDPD
LPALQPHGYDQPGTEPLDLHNNGHTVQLSLPSTLYLGGLPRKYVAAQLHLHWGQKGSPGG
SEHQINSEATFAELHIVHYDSDSYDSLSEAAERPQGLAVLGILIEVGETKNIAYEHILSH
LHEVRHKDQKTSVPPFNLRELLPKQLGQYFRYNGSLTTPPCYQSVLWTVFYRRSQISMEQ
LEKLQGTLFSTEEEPSKLLVQNYRALQPLNQRMVFASFIQAGSSYTTGEMLSLGVGILVG
CLCLLLAVYFIARKIRKKRLENRKSVVFTSAQATTEA


We've got our sequence above but there's a small problem - it's split across 6 lines. Ideally, we'd like our sequence to be one continuous string so we can easily compare it to other sequences.

In [21]:
# open the FASTA file in 'read' mode
fasta_file = open('fasta/small.fasta','r').readlines()

# create an 'empty' string using the '' or "" quote marks
this_protein = ''

# iterate through lines
for line in fasta_file:
    if '>' not in line:
        
        # here, we add the new line to our empty string
        this_protein += line.strip()

# print the result
print(this_protein)

MLFSALLLEVIWILAADGGQHWTYEGPHGQDHWPASYPECGNNAQSPIDIQTDSVTFDPDLPALQPHGYDQPGTEPLDLHNNGHTVQLSLPSTLYLGGLPRKYVAAQLHLHWGQKGSPGGSEHQINSEATFAELHIVHYDSDSYDSLSEAAERPQGLAVLGILIEVGETKNIAYEHILSHLHEVRHKDQKTSVPPFNLRELLPKQLGQYFRYNGSLTTPPCYQSVLWTVFYRRSQISMEQLEKLQGTLFSTEEEPSKLLVQNYRALQPLNQRMVFASFIQAGSSYTTGEMLSLGVGILVGCLCLLLAVYFIARKIRKKRLENRKSVVFTSAQATTEA


OK, so we know how to assemble the whole protein sequence into one string. So far, this works fine for our small fasta file (which only has one protein), but what if the file had more?

Let's write some code that uses the ```medium.fasta``` file as input.

First, we'll count the number of proteins:

In [24]:
# open the FASTA file in 'read' mode
fasta_file = open('fasta/medium.fasta','r').readlines()

counter = 0

# iterate through lines
for line in fasta_file:
    
    # add 1 to our counter for every header line we see
    if '>' in line:
        counter += 1

# print the result
print(counter)

2


In [31]:
# open the FASTA file in 'read' mode
fasta_file = open('fasta/medium.fasta','r').readlines()

# create an 'empty' list to store the sequences
sequences = []

# iterate through lines
for line in fasta_file:
    
    if '>' in line:
        # for every header line we see, add an empty string to the list
        # -- this empty string will later be populated with the sequence
        sequences.append('')
        
    if '>' not in line:
        # here, we add the new line to our empty string
        # note that sequences[-1] refers to the last item in the list
        sequences[-1] += line.strip()

# print the result
for sequence in sequences:
    print(sequence)

MLFSALLLEVIWILAADGGQHWTYEGPHGQDHWPASYPECGNNAQSPIDIQTDSVTFDPDLPALQPHGYDQPGTEPLDLHNNGHTVQLSLPSTLYLGGLPRKYVAAQLHLHWGQKGSPGGSEHQINSEATFAELHIVHYDSDSYDSLSEAAERPQGLAVLGILIEVGETKNIAYEHILSHLHEVRHKDQKTSVPPFNLRELLPKQLGQYFRYNGSLTTPPCYQSVLWTVFYRRSQISMEQLEKLQGTLFSTEEEPSKLLVQNYRALQPLNQRMVFASFIQAGSSYTTGEMLSLGVGILVGCLCLLLAVYFIARKIRKKRLENRKSVVFTSAQATTEA
MPPSISAFQAAYIGIEVLIALVSVPGNVLVIWAVKVNQALRDATFCFIVSLAVADVAVGALVIPLAILINIGPQTYFHTCLMVACPVLILTQSSILALLAIAVDRYLRVKIPLRYKMVVTPRRAAVAIAGCWILSFVVGLTPMFGWNNLSAVERAWAANGSMGEPVIKCEFEKVISMEYMVYFNFFVWVLPPLLLMVLIYLEVFYLIRKQLNKKVSASSGDPQKYYGKELKIAKSLALILFLFALSWLPLHILNCITLFCPSCHKPSILTYIAIFLTHGNSAMNPIVYAFRIQKFRVTFLKIWNDHFRCQPAPPIDEDLPEERPDD


## Finding mutations in protein sequences between two subjects

Ok, let's move on to something a bit more involved. Let's say that we've got a list of protein sequences found in two different people and we suspect that one person has a detrimental mutation in at least one protein. Our task is to read the files containing the protein sequences for each person, iterate through them, and compare their sequences to find any differences.

The protein sequences are in the ```fasta``` directory and are labelled ```subject_1.fasta``` and ```subject_2.fasta```

#### Exploring the data
Let's check how many proteins we have in each file

In [17]:
def count_proteins(file_name):

    # create an empty list
    # here, we'll store the FASTA header lines as we read them from the files
    headers = []
    
    # open the FASTA file in 'read' mode
    fasta_file = open(file_name,'r').readlines()
    
    # iterate through lines
    for line in fasta_file:
        
        # we're interested in the header lines, which start with '>'
        if '>' in line:
            
            # we've found a header line
            # - append to the list
            headers.append(line)
    
    # return the list from the function
    return headers

# read the subject_1 FASTA file
subject_1_headers = count_proteins('fasta/subject_1.fasta')

# read the subject_2 FASTA file
subject_2_headers = count_proteins('fasta/subject_2.fasta')

print(len(subject_1_headers))
print(len(subject_2_headers))

print(subject_1_headers == subject_2_headers)

3684
3684
True


Hold on! We just did something new! What's this ```def count_proteins(file_name):``` business?

### Functions
Above we wanted to open two different files, and collect the header lines from both. We already knew how to do that from our work above. For example, we could have written something like:

In [16]:
subject_1_headers = []
fasta_file_1 = open('fasta/subject_1.fasta','r').readlines()
for line in fasta_file_1:
    if '>' in line:
        subject_1_headers.append(line)

subject_2_headers = []
fasta_file_2 = open('fasta/subject_2.fasta','r').readlines()
for line in fasta_file_2:
    if '>' in line:
        subject_2_headers.append(line)
        
print(len(subject_1_headers))
print(len(subject_2_headers))

3684
3684


This gets us to the exact same point. It's fine but the problem is that we're repeating a lot of code. The lines that open the files, iterate through the contents, filter for header (```>```) lines and append to a list are the same for both!

Generally, we want to write as little code as possible. This helps to keep things easy to maintain, makes mistakes easier to find and fix, and just generally keeps things nice!

In our first code block, we're written what we call a 'function'.

Functions are blocks of code that we write once and then can re-use with different input ```arguments```.

A more simple example is below. Here, we define (```def```) that we'll call ```multiply_by_two```. This function takes one input (```x```) that we work with inside the function and then ```return``` back to the line of code that asked for it.

In [12]:
# define a new function with 'def'
def multiply_by_two(x):
    result = x * 2
    return result

print(multiply_by_two(1))
print(multiply_by_two(10))
print(multiply_by_two(5))

2
20
10


Functions can have as many input ```arguments``` as we need:

In [13]:
def multiply_x_by_y(x, y):
    result = x * y
    return result

print(multiply_x_by_y(1, 100))
print(multiply_x_by_y(10, 2))
print(multiply_x_by_y(5, 3))

100
20
15


Ok, back to our fasta files. We know that we have the same number of proteins in each file because the number of header lines is the same. 

Let's now write a function to read the sequences.

In [14]:
def read_protein_sequences(file_name):
    
    # open the FASTA file in 'read' mode
    fasta_file = open(file_name,'r').readlines()

    # create an 'empty' list to store the sequences
    sequences = []

    # iterate through lines
    for line in fasta_file:

        if '>' in line:
            # for every header line we see, add an empty string to the list
            # -- this empty string will later be populated with the sequence
            sequences.append('')

        if '>' not in line:
            # here, we add the new line to our empty string
            # note that sequences[-1] refers to the last item in the list (the item most recently added)
            sequences[-1] += line.strip()
    
    return sequences

# read the subject_1 FASTA file
subject_1_sequences = read_protein_sequences('fasta/subject_1.fasta')

# read the subject_2 FASTA file
subject_2_sequences = read_protein_sequences('fasta/subject_2.fasta')

Now we've got our protein sequences from both subjects. Let's see if there are any differences:

In [41]:
print(subject_1_sequences == subject_2_sequences)

False


There are some differences between our subjects! Right now, we don't know how many proteins or amino acids differ though. Lets find them:

In [48]:
for i in range(len(subject_1_sequences)):

    # is the i'th sequence from subject 1 different to the i'th sequence from subject 2?
    # -- print the sequences if they are different
    if subject_1_sequences[i] != subject_2_sequences[i]:
        print(subject_1_sequences[i])
        print(subject_2_sequences[i])

MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGPDEAPRMPEAAPPVAPAPAAPTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAKSVTCTYSPALNKMFCQLAKTCPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHERCSDSDGLAPPQHLIRVEGNLRVEYLDDRNTFRHSVVVPYEPPEVGSDCTTIHYNYMCNSSCMGGMNRRPILTIITLEDSSGNLLGRNSFEVRVCACPGRDRRTEEENLRKKGEPHHELPPGSTKRALPNNTSSSPQPKKKPLDGEYFTLQIRGRERFEMFRELNEALELKDAQAGKEPGGSRAHSSHLKSKKGQSTSRHKKLMFKTEGPDSD
MEEPQSDPSVEPPLSQETFSDLWKLLPENNVLSPLPSQAMDDLMLSPDDIEQWFTEDPGPDEAPRMPEAAPPVAPAPAAPTPAAPAPAPSWPLSSSVPSQKTYQGSYGFRLGFLHSGTAKSVTCTYSPALNKMFCQLAKTCPVQLWVDSTPPPGTRVRAMAIYKQSQHMTEVVRRCPHHERCSDSDGLAPPQHLIRVEGNLRVEYLDDRNTFRHSVVVPYEPPEVGSDCTTIHYNYMCNSSCMGGMNRRPILTIITLEDSSGNLLGRNSFEVHVCACPGRDRRTEEENLRKKGEPHHELPPGSTKRALPNNTSSSPQPKKKPLDGEYFTLQIRGRERFEMFRELNEALELKDAQAGKEPGGSRAHSSHLKSKKGQSTSRHKKLMFKTEGPDSD


We did something new here. ```range(len(subject_1_sequences))```... what's that?

We need a way of retrieving the first, second, third.... elements of the ```subject_1_sequences``` list and the first element of the ```subject_2_sequences``` lists so that we can compare them.

Remember that we can retrieve elements from a list ```x``` with the ```x[0]```, ```x[1]```, ```x[2]```... syntax. 

```range()``` will create a list* of integers of a given length starting from 0.

For example, ```range(5)``` will create ```[0,1,2,3,4]``` and ```range(10)``` will create ```[0,1,2,3,4,5,6,7,8,9]```.

When we call ```range(len(subject_1_sequences))```, we're saying make a list of integers starting at 0 of the same length as the list ```subject_1_sequences```. This is handy since we can use these integers to retrieve the relevant elements from our sequence lists.





*range actually creates a generator, but that's for another time

Can you spot the amino acid that's different? It's a bit hard. Let's write some code to find it for us:

In [18]:
for i in range(len(subject_1_sequences)):
    
    # is the i'th sequence from subject 1 different to the i'th sequence from subject 2?
    if subject_1_sequences[i] != subject_2_sequences[i]:
        
        # search through the sequences of both proteins 
        for j in range(len(subject_1_sequences[i])):
            
            # is residue j of protein 1 different to residue [j] of protein 2? 
            if subject_1_sequences[i][j] != subject_2_sequences[i][j]:
                
                # if so, print the result
                print('%s: %s --> %s' % (
                    j+1, subject_1_sequences[i][j], subject_2_sequences[i][j])
                )
                
                # print the fasta header line as sell
                print(subject_1_headers[i])

273: R --> H
>sp|P04637|P53_HUMAN Cellular tumor antigen p53 OS=Homo sapiens OX=9606 GN=TP53 PE=1 SV=4



Residue arginine 273 of the p53 protein has been mutated to histidine. Biologically speaking that's an important mutation and is known to be involved in the genesis of many cancers.

## Thanks!

We hope you've enjoyed this tutorial and find it useful! For questions, contact:

- Dr. Nikeisha Caruana (nikeisha.caruana@unimelb.edu.au)
- Dr. Michael Leeming (leemingm@unimelb.edu.au)